src/group_theory/finiteness.lean

```diff
@@ -101,6 +102,45 @@ add_monoid.fg_iff_mul_fg.1 ‹_›
 
 end monoid
 
+namespace submonoid
+
+@[to_additive]
+lemma fg_iff_monoid_fg [decidable_eq M] (N : submonoid M) : N.fg ↔ monoid.fg N :=
+begin
+  split, swap,
+  { exact λ s, ⟨s.1.some.image N.subtype, by rw [finset.coe_image, ← monoid_hom.map_mclosure,
+    s.1.some_spec, ← monoid_hom.mrange_eq_map, range_subtype]⟩ },
+  rintro ⟨s, hs⟩,
+  let s' := s.preimage N.subtype (by tidy),
+  have : (s'.image N.subtype : set M) = s := by simpa [← hs, -subset_closure] using subset_closure,
+  refine ⟨⟨s', _⟩⟩,
+  rw eq_top_iff,
+  rintros ⟨x, hx⟩ -,
+  rw [← hs, ← this, finset.coe_image, ← monoid_hom.map_mclosure] at hx,
+  rcases hx with ⟨x, hx', rfl⟩,
+  simp only [set_like.eta, submonoid.coe_subtype],
+  exact hx'
+end
+
+@[to_additive]
+instance [decidable_eq M] (r : M) : monoid.fg (submonoid.powers r) :=
+(fg_iff_monoid_fg _).mp ⟨({r} : finset M), by simpa using (powers_eq_closure _).symm⟩
+
+@[to_additive]
+lemma _root_.monoid.fg_of_surjective {M' : Type*} [monoid M'] [decidable_eq M'] [h : monoid.fg M]
+  (f : M →* M') (hf : function.surjective f) :
+  monoid.fg M' :=
+begin
+  tactic.unfreeze_local_instances,
+  rcases h with ⟨s, hs⟩,
```

Comment:
This can't be the idiomatic way to do this... I'll come back tomorrow and look for a nicer way

Reply:
I think `obtain ⟨s, hs⟩ := monoid.fg_def.1 ‹_›` is slightly better, and then you don't have to name the typeclass or add the `unfreeze_local_instances`